server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,205 @@
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 
+import javax.mail.Header;
 import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
 
 import org.apache.james.core.MailAddress;
+import org.apache.james.mime4j.dom.Message;
+import org.apache.james.mime4j.stream.MimeConfig;
+import org.apache.james.util.mime.MessageContentExtractor;
+import org.apache.james.util.mime.MessageContentExtractor.MessageContent;
+import org.apache.james.util.streams.Iterators;
 import org.apache.mailet.Mail;
+import org.apache.mailet.PerRecipientHeaders;
 
+import com.github.fge.lambdas.Throwing;
 import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
 
 public class MailDto {
-    public static MailDto fromMail(Mail mail) throws MessagingException {
+    public static MailDto fromMail(Mail mail, List<AdditionalField> additionalFields) throws MessagingException, InnaccessibleFieldException {
         return new MailDto(mail.getName(),
             Optional.ofNullable(mail.getSender()).map(MailAddress::asString),
             mail.getRecipients().stream().map(MailAddress::asString).collect(Guavate.toImmutableList()),
             Optional.ofNullable(mail.getErrorMessage()),
-            Optional.ofNullable(mail.getState()));
+            Optional.ofNullable(mail.getState()),
+            Optional.ofNullable(mail.getRemoteHost()),
+            Optional.ofNullable(mail.getRemoteAddr()),
+            Optional.ofNullable(mail.getLastUpdated()),
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(mail.getMessageSize());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.MESSAGE_SIZE, e);
+        }
+    }
+
+    private static Optional<String> fetchBody(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
```

Comment:
we build a Dto to not transfer it ? Didn't we abuse it because it has a handy feature we want ?

Reply:
I'm not sure to get it, can you re-explain it please?

Reply:
Yes @mbaechler  there is a mis-understanding here: `convertMessage` does convert the message to MIME4J message. It do not build a DTO.